packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+import { Rate, RateLiteral, parseRate } from './rate'
+import { Slot, SlotProvider } from './slot'
+import { types } from '@mondrian-framework/model'
+import { functions } from '@mondrian-framework/module'
+
+//inspired by https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+/**
+ * Rate limit at `rateLimit` in the `samplingPeriod`.
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
+    this.rate = typeof rate === 'string' ? parseRate(rate) : rate
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
+    const slidingWindow = this.slidingWindows.get(key)
+    if (slidingWindow) {
+      return slidingWindow
+    }
+    const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
+    this.slidingWindows.set(key, newSlidingWindow)
+    return newSlidingWindow
+  }
+}
+
+export class SlidingWindow {
+  private readonly samplingPeriodSeconds: number
+  private readonly rateLimit: number
+  private readonly slots: Map<number, Slot> = new Map()
+  private readonly nowSeconds: () => number
+  private readonly slotProvider: SlotProvider
+  private readonly key: string
+  private rateLimitedUntil?: number
+
+  constructor({
+    rate,
+    nowSeconds,
+    slotProvider,
+    key,
+  }: {
+    rate: Rate
+    nowSeconds?: () => number
+    slotProvider: SlotProvider
+    key: string
+  }) {
```

Comment:
The inlined type annotation makes the code harder to read, I'd move it into its own type

Also it would be useful to have documentation that explains what each argument does:
- what is the rate used for?
- what is `nowSeconds`? Why is it a function? What is it used for? What happens if it's undefined? What happens if the function throws? Does it really need to be a function?
- what is key? Why does it need it?